### PR TITLE
[SPARK-14852][ML] refactored GLM summary into training, non-training summaries

### PR DESCRIPTION
**Pull request by jkbradley**
## What changes were proposed in this pull request?

This splits GeneralizedLinearRegressionSummary into 2 summary types:
- GeneralizedLinearRegressionSummary, which does not store info from fitting (diagInvAtWA)
- GeneralizedLinearRegressionTrainingSummary, which is a subclass of GeneralizedLinearRegressionSummary and stores info from fitting

This also add a method evaluate() which can produce a GeneralizedLinearRegressionSummary on a new dataset.

The summary no longer provides the model itself as a public val.

Also:
- Fixes bug where GeneralizedLinearRegressionTrainingSummary was created with model, not summaryModel.
- Adds hasSummary method.
- Renames findSummaryModelAndPredictionCol -> getSummaryModel and simplifies that method.
- In summary, extract values from model immediately in case user later changes those (e.g., predictionCol).
- Pardon the style fixes; that is IntelliJ being obnoxious.
## How was this patch tested?

Existing unit tests + updated test for evaluate and hasSummary

### CODE REVIEW

**Comment by jkbradley** [comment]
CC: @yanboliang Would you have time to take a look at this?  I'd like to get this into 2.0 (before this becomes a public API).  Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala [hunk 1]
```diff
@@ -761,16 +783,24 @@ class GeneralizedLinearRegressionModel private[ml] (
...
-  private[regression] def findSummaryModelAndPredictionCol()
-    : (GeneralizedLinearRegressionModel, String) = {
+  private[regression] def getSummaryModel: GeneralizedLinearRegressionModel = {
     $(predictionCol) match {
       case "" =>
         val predictionColName = "prediction_" + java.util.UUID.randomUUID.toString
-        (copy(ParamMap.empty).setPredictionCol(predictionColName), predictionColName)
-      case p => (this, p)
+        copy(ParamMap.empty).setPredictionCol(predictionColName)
+      case p => this
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala]: I'd like to change it to ```case p => copy(ParamMap.empty)```, and this will make ```GeneralizedLinearRegressionTrainingSummary``` always along with an immutable model.
Users want to learn about the training summary is a specific model rather than an updated model. If model was updated later not by retrain, we should not provide updated training summary.

File: mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala [hunk 2]
```diff
@@ -834,43 +864,51 @@ object GeneralizedLinearRegressionModel extends MLReadable[GeneralizedLinearRegr
...
+  private[regression] val familyName: String = model.getFamily
+  private[regression] val family: Family = Family.fromName(familyName)
+  private[regression] val link: Link = if (model.isDefined(model.link)) {
     Link.fromName(model.getLink)
   } else {
     family.defaultLink
   }
+  private[regression] val fitIntercept: Boolean = model.getFitIntercept
+  private[regression] val intercept: Double = model.intercept
+  private[regression] val coefficients: Vector = model.coefficients
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala]: If the model of ```GeneralizedLinearRegressionSummary``` is immutable, it's not necessary to extract these values in advance.

**Comment by yanboliang** [comment]
@jkbradley This looks good overall, only one concern at inline comment. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala (hunk 1 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala]: That sounds good.  That way, we can just use the model internally in the summary and will not need to copy the model parameters below.  I'll do that.

**Comment by jkbradley** [comment]
Thanks for the suggestion; this simplifies things.  I just pushed an update, but please hold off on review while I make a pass over the diff.

**Comment by jkbradley** [comment]
OK, now ready for another review pass.

File: mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala [hunk 3]
```diff
@@ -834,36 +836,55 @@ object GeneralizedLinearRegressionModel extends MLReadable[GeneralizedLinearRegr
...
+    if (origModel.isDefined(origModel.predictionCol) && origModel.getPredictionCol != "") {
+      origModel.getPredictionCol
+    } else {
+      "prediction_" + java.util.UUID.randomUUID.toString
+    }
+  }
+
+  /**
+   * Private copy of model to ensure Params are not modified outside this class.
+   * Coefficients is not a deep copy, but that is acceptable.
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala]: L874 will deep copy the model including coefficients and intercept, so I think the document is out of date.

**Comment by yanboliang** [comment]
LGTM except the last minor issue. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala (hunk 3 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala]: I'm pretty sure it's a shallow copy of the coefficients: [https://github.com/apache/spark/blob/89addd40abdacd65cc03ac8aa5f9cf3dd4a4c19b/mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala#L776]

**Comment by jkbradley** [comment]
Thanks for reviewing!  I'll merge this with master
(But if I'm wrong about that last comment, I can do a follow-up fix.)

**Comment by yanboliang** [comment]
Oh, sorry for misunderstand. You are right, thanks for remind.
